omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,899 @@
+
+    // Used for synchronization of (pseudo)remaining work.
+    int threads_working = 0;
+
+#pragma omp parallel firstprivate(max_degree)
+    {
+        // This is an upper bound, as '.deque_chunk...' is upper bounded by
+        // chunk_bound.
+        gko::vector<IndexType> local_to_insert(max_degree * chunk_bound, exec);
+        auto local_to_insert_size = 0;
+
+        while (true) {
+            // Get some nodes from the locked q.
+            const auto pair = q.dequeue_chunk(&threads_working);
+            const auto chunk = pair.first;
+            const auto chunk_size = pair.second;
+
+            // If there are nodes left to process.
+            // If there are no nodes returned that also means there are no nodes
+            // still in flight.
+            if (chunk_size > 0) {
+                // For each node in the chunk.
+                for (auto chunk_i = 0; chunk_i < chunk_size; ++chunk_i) {
+                    const auto node = chunk[chunk_i];
+
+                    IndexType level;
+#pragma omp atomic read
+                    level = levels[node];
+
+                    const auto new_neighbour_level = level + 1;
+
+                    // Go through neighbours.
+                    const auto row_start = row_ptrs[node];
+                    const auto row_end = row_ptrs[node + 1];
+
+                    for (auto neighbour_i = row_start; neighbour_i < row_end;
+                         ++neighbour_i) {
+                        const auto neighbour = col_idxs[neighbour_i];
+
+                        IndexType old_neighbour_level;
+                        do {
+#pragma omp atomic read
+                            old_neighbour_level = levels[neighbour];
+
+                            const auto is_new_smaller =
+                                new_neighbour_level < old_neighbour_level;
+
+                            if (!is_new_smaller) {
+                                goto continue_outer;
+                            }
+                        } while (!__atomic_compare_exchange_n(
+                            &levels[neighbour], &old_neighbour_level,
+                            new_neighbour_level, true,
+                            std::memory_order::memory_order_acq_rel,
+                            std::memory_order::memory_order_acquire));
+
+                        local_to_insert[local_to_insert_size++] = neighbour;
+
+                    continue_outer:;
+                    }
+                }
+
+                // Insert back to global work queue, done with this chunk.
+                q.enqueue_chunk(&local_to_insert[0], local_to_insert_size);
+
+#pragma omp atomic update
+                --threads_working;
+
+                local_to_insert_size = 0;
+            } else {
+                // Done!
+                break;
+            }
+        }
+    }
+}
+
+
+/**
+ * Finds a 'contender', meaning a node in the last level of the rls with minimum
+ * degree, returns it along with the rls height.
+ */
```

Comment:
I thought about that extensively. However, ubfs is claimed to scale relatively well, meaning the runtime advantage running two searches with half threads each as compared to two searches serially with all threads each should be small.

That's not even considering the most important aspect: We cant look at a starting node and claim: "This one is better than the other one!". There is the _tendency_ of pseudoperipheral nodes to perform well, probably because their rls have great heights, however I remember reading a quote by Alan George, which basically said that even if we could find peripheral nodes easily, there is no reason to believe they would be any better suited than pseudoperipheral nodes, even though their rls have a (slightly) greater height.

If I were to implement this anyway, I would choose the n nodes with smallest degree. There is, in general, no reason to believe they shouldn't be random across the graph.